File: lachancuocgoi/app/src/main/java/com/example/lachancuocgoi/Analysis/L2/WFSA/WfsaEngine.kt
package com.example.lachancuocgoi.Analysis.L2.WFSA

import android.util.Log
import com.example.lachancuocgoi.Analysis.L2.GDetection.GFlash
import com.example.lachancuocgoi.Analysis.L2.Intent.IntentPrediction
import com.example.lachancuocgoi.Analysis.L2.Intent.ScamIntent

enum class ScamStage(val weightMultiplier: Double) {
    STAGE_1_INTRODUCTION(1.0),
    STAGE_2_BAITING_THREAT(1.5),
    STAGE_3_URGENCY(2.0),
    STAGE_4_COMMAND(3.0)
}

data class StateNode(
    val id: String,
    val description: String,
    val stage: ScamStage
)

data class Transition(
    val triggerPhrases: List<String>,
    val targetStateId: String,
    val requiredIntent: ScamIntent? = null
) {
    /**
     * [A2] Pre-tokenized trigger phrases using GFlash for consistent matching.
     * Lazy initialized on first access
     */
    val normalizedTriggerSets: List<Set<String>> by lazy {
        triggerPhrases.map { phrase ->
            GFlash.tokenize(phrase).toSet()
        }.filter { it.isNotEmpty() }
    }
}

data class ScenarioGraph(
    val graphId: String,
    val name: String,
    val states: Map<String, StateNode>,
    val transitions: Map<String, List<Transition>>,
    val initialStateId: String
)

class WfsaEngine(private val graphs: List<ScenarioGraph>) {
    private val TAG = "WfsaEngine"
    
    // [C1] Per-graph scoring thay vì global accumulation
    private val graphScores = mutableMapOf<String, Double>()
    private val currentSessionStates = mutableMapOf<String, String>()

    // Decay: mỗi segment, score cũ nhân với decayFactor (< 1.0) để giảm ảnh hưởng trigger cũ
    private val segmentCountThreshold = 3  // Sau 3 segment không có trigger mới, bắt đầu decay nhanh hơn
    private val perGraphSegmentsSinceLastTrigger = mutableMapOf<String, Int>()

    // [C1] Track active scenario for UI reporting
    var activeScenarioName: String? = null
        private set
    var activeScenarioStage: Int? = null
        private set

    init {
        resetSession()
    }

    fun resetSession() {
        currentSessionStates.clear()
        graphScores.clear()
        perGraphSegmentsSinceLastTrigger.clear()
        graphs.forEach { graph ->
            currentSessionStates[graph.graphId] = graph.initialStateId
            graphScores[graph.graphId] = 0.0
            perGraphSegmentsSinceLastTrigger[graph.graphId] = 0
        }
        activeScenarioName = null
        activeScenarioStage = null
        Log.i(TAG, "WFSA Session Reset. Tracking ${graphs.size} scenarios.")
    }

    /**
     * [C2] Adaptive decay — kịch bản nguy hiểm decay chậm hơn.
     */
    private fun getDecayForGraph(graphId: String): Double {
        return when {
            graphId.startsWith("G_POLICE") || graphId.startsWith("G_KIDNAP") || 
            graphId.startsWith("G_SEXTORT") -> 0.95 // Slow decay cho kịch bản nguy hiểm nhất
            graphId.startsWith("G_CHARITY") || graphId.startsWith("G_GENERIC") || 
            graphId.startsWith("G_GAMBLE") -> 0.80 // Fast decay cho kịch bản ít nguy hiểm
            else -> 0.90 // Default
        }
    }

    /**
     * [A2+C1+C2] Phân tích đoạn văn bản mới dựa trên đồ thị trạng thái và ý định (Intents).
     * 
     * Cải tiến:
     * - A2: Dùng GFlash tokenization thay vì raw string contains
     * - C1: Per-graph scoring (MAX) thay vì global accumulation (SUM)
     * - C2: Adaptive decay theo mức nguy hiểm của scenario
     *
     * @param newTranscript Đoạn văn bản mới được thêm vào từ stream.
     * @param intentPredictions Các ý định nhận diện được từ TFLite (GĐ2).
     * @return Điểm rủi ro tích lũy hiện tại (Risk Score) = MAX of all graphs.
     */
    fun analyzeSegment(newTranscript: String, intentPredictions: List<IntentPrediction>): Double {
        // [A2] Dùng GFlash tokenization thay vì raw lowercase
        val normalizedTokens = GFlash.tokenize(newTranscript).toSet()
        if (normalizedTokens.isEmpty()) {
            return getCurrentRiskScore()
        }

        // Trích xuất intent có độ tin cậy cao (chuẩn bị cho GĐ3 tích hợp)
        val dominantIntents = intentPredictions.filter { it.confidence > 0.6f }.map { it.intent }

        // [C1] Per-graph decay + scoring
        graphs.forEach { graph ->
            val graphId = graph.graphId
            val currentScore = graphScores[graphId] ?: 0.0
            val segmentsSinceLastTrigger = perGraphSegmentsSinceLastTrigger[graphId] ?: 0

            // Decay score cũ trước khi xử lý segment mới
            if (currentScore > 0) {
                val baseDecay = getDecayForGraph(graphId)
                val effectiveDecay = if (segmentsSinceLastTrigger >= segmentCountThreshold) {
                    baseDecay * 0.85  // Decay nhanh hơn khi lâu không có trigger
                } else {
                    baseDecay
                }
                val newScore = currentScore * effectiveDecay
                graphScores[graphId] = if (newScore < 1.0) 0.0 else newScore
            }
            perGraphSegmentsSinceLastTrigger[graphId] = segmentsSinceLastTrigger + 1

            // Check transitions
            val currentStateId = currentSessionStates[graphId] ?: graph.initialStateId
            val possibleTransitions = graph.transitions[currentStateId] ?: emptyList()

            for (transition in possibleTransitions) {
                // [A2] Dùng pre-tokenized trigger sets thay vì raw contains
                val hasPhraseMatch = transition.normalizedTriggerSets.any { triggerTokens ->
                    triggerTokens.all { it in normalizedTokens }
                }
                
                val hasIntentMatch = transition.requiredIntent != null && dominantIntents.contains(transition.requiredIntent)

                // Điều kiện chuyển trạng thái: Có từ khóa HOẶC đúng loại ý định (Intent)
                if (hasPhraseMatch || hasIntentMatch) {
                    currentSessionStates[graphId] = transition.targetStateId
                    val targetState = graph.states[transition.targetStateId]
                    
                    if (targetState != null) {
                        val triggerType = if (hasIntentMatch) "Intent:${transition.requiredIntent}" else "Phrase"
                        Log.w(TAG, "WFSA Transition [$triggerType]: $currentStateId -> ${targetState.id} (Stage: ${targetState.stage.name}) [${graph.name}]")
                        val scoreIncrease = 10.0 * targetState.stage.weightMultiplier
                        graphScores[graphId] = ((graphScores[graphId] ?: 0.0) + scoreIncrease).coerceAtMost(100.0)
                        perGraphSegmentsSinceLastTrigger[graphId] = 0  // Reset counter khi có trigger mới
                    }
                    break
                }
            }
        }

        // [C1] Update active scenario info for UI
        val bestGraph = graphScores.maxByOrNull { it.value }
        if (bestGraph != null && bestGraph.value > 0) {
            val graph = graphs.find { it.graphId == bestGraph.key }
            activeScenarioName = graph?.name
            val currentState = currentSessionStates[bestGraph.key]
            val stateNode = graph?.states?.get(currentState)
            activeScenarioStage = stateNode?.stage?.ordinal?.plus(1)
        } else {
            activeScenarioName = null
            activeScenarioStage = null
        }

        val maxScore = getCurrentRiskScore()
        if (maxScore > 0) {
            Log.d(TAG, "Current Max Risk Score: $maxScore (from ${bestGraph?.key ?: "none"})")
        }

        return maxScore
    }
    
    // [C1] MAX of all graph scores thay vì single accumulated score
    fun getCurrentRiskScore(): Double = graphScores.values.maxOrNull() ?: 0.0
}

File: lachancuoicgoi_flutter/lib/analysis/l2/wfsa/wfsa_engine.dart
import '../g_detection/g_flash.dart';
import '../intent/scam_intent.dart';

enum ScamStage {
  stage1Introduction(1.0),
  stage2BaitingThreat(1.5),
  stage3Urgency(2.0),
  stage4Command(3.0);

  const ScamStage(this.weightMultiplier);

  final double weightMultiplier;
}

class StateNode {
  const StateNode({
    required this.id,
    required this.description,
    required this.stage,
  });

  final String id;
  final String description;
  final ScamStage stage;
}

class Transition {
  Transition({
    required this.triggerPhrases,
    required this.targetStateId,
    this.requiredIntent,
  });

  final List<String> triggerPhrases;
  final String targetStateId;
  final ScamIntent? requiredIntent;

  late final List<Set<String>> normalizedTriggerSets = triggerPhrases
      .map((phrase) => GFlash.tokenize(phrase).toSet())
      .where((tokens) => tokens.isNotEmpty)
      .toList();
}

class ScenarioGraph {
  const ScenarioGraph({
    required this.graphId,
    required this.name,
    required this.states,
    required this.transitions,
    required this.initialStateId,
  });

  final String graphId;
  final String name;
  final Map<String, StateNode> states;
  final Map<String, List<Transition>> transitions;
  final String initialStateId;
}

/// Records a single state transition for timeline analysis.
class TransitionRecord {
  const TransitionRecord({
    required this.graphId,
    required this.fromStateId,
    required this.toStateId,
    required this.scoreAfter,
    required this.stage,
    required this.segmentIndex,
  });

  final String graphId;
  final String fromStateId;
  final String toStateId;
  final double scoreAfter;
  final ScamStage stage;
  final int segmentIndex;
}

class WfsaEngine {
  WfsaEngine(this.graphs) {
    resetSession();
  }

  final List<ScenarioGraph> graphs;
  final Map<String, double> _graphScores = <String, double>{};
  final Map<String, String> _currentSessionStates = <String, String>{};
  final Map<String, int> _segmentsSinceLastTrigger = <String, int>{};

  /// Session timeline: ordered list of all transitions across all graphs.
  final List<TransitionRecord> _transitionHistory = [];

  /// Read-only access to the transition timeline for analysis.
  List<TransitionRecord> get transitionHistory =>
      List<TransitionRecord>.unmodifiable(_transitionHistory);

  /// Monotonically increasing segment counter for timeline tracking.
  int _segmentIndex = 0;

  /// Tracks how many characters of the transcript have been analyzed.
  int _analyzedTextLength = 0;

  String? activeScenarioName;
  int? activeScenarioStage;

  void resetSession() {
    _currentSessionStates.clear();
    _graphScores.clear();
    _segmentsSinceLastTrigger.clear();
    _transitionHistory.clear();
    _segmentIndex = 0;
    _analyzedTextLength = 0;
    for (final graph in graphs) {
      _currentSessionStates[graph.graphId] = graph.initialStateId;
      _graphScores[graph.graphId] = 0;
      _segmentsSinceLastTrigger[graph.graphId] = 0;
    }
    activeScenarioName = null;
    activeScenarioStage = null;
  }

  /// Incremental analysis: only tokenize the delta portion of [fullText]
  /// that hasn't been analyzed yet.
  double analyzeIncremental(
    String fullText,
    List<IntentPrediction> intentPredictions,
  ) {
    if (fullText.length < _analyzedTextLength) {
      _analyzedTextLength = 0;
    }
    if (fullText.length <= _analyzedTextLength) {
      return currentRiskScore;
    }
    final deltaText = fullText.substring(_analyzedTextLength);
    _analyzedTextLength = fullText.length;
    return analyzeSegment(deltaText, intentPredictions);
  }

  double analyzeSegment(
    String newTranscript,
    List<IntentPrediction> intentPredictions,
  ) {
    final normalizedTokens = GFlash.tokenize(newTranscript).toSet();
    if (normalizedTokens.isEmpty) return currentRiskScore;

    _segmentIndex++;

    final dominantIntents = intentPredictions
        .where((prediction) => prediction.confidence > 0.6)
        .map((prediction) => prediction.intent)
        .toSet();

    for (final graph in graphs) {
      final graphId = graph.graphId;
      final currentScore = _graphScores[graphId] ?? 0;
      final segmentsSinceLast = _segmentsSinceLastTrigger[graphId] ?? 0;

      // --- Adaptive decay: slower at higher stages (scammer deep in scenario) ---
      if (currentScore > 0) {
        final currentStateId =
            _currentSessionStates[graphId] ?? graph.initialStateId;
        final currentState = graph.states[currentStateId];
        final currentStage = currentState?.stage ?? ScamStage.stage1Introduction;
        final baseDecay = _decayForGraph(graphId);
        final adaptiveDecay = _adaptiveDecay(baseDecay, currentStage);
        final newScore = currentScore * adaptiveDecay;
        _graphScores[graphId] = newScore < _minScoreFloor ? 0.0 : newScore;
      }
      _segmentsSinceLastTrigger[graphId] = segmentsSinceLast + 1;

      final currentStateId =
          _currentSessionStates[graphId] ?? graph.initialStateId;
      final possibleTransitions =
          graph.transitions[currentStateId] ?? const <Transition>[];

      for (final transition in possibleTransitions) {
        final hasPhraseMatch = transition.normalizedTriggerSets.any(
          (triggerTokens) =>
              triggerTokens.every((token) => normalizedTokens.contains(token)),
        );
        final requiredIntent = transition.requiredIntent;
        final hasIntentMatch =
            requiredIntent != null && dominantIntents.contains(requiredIntent);

        if (hasPhraseMatch || hasIntentMatch) {
          final fromStateId = currentStateId;
          _currentSessionStates[graphId] = transition.targetStateId;
          final targetState = graph.states[transition.targetStateId];
          if (targetState != null) {
            var scoreIncrease = 10.0 * targetState.stage.weightMultiplier;

            // --- Intent-boost: double score when TFLite intent confirms ---
            if (hasIntentMatch && hasPhraseMatch) {
              scoreIncrease *= _intentBoostFactor;
            }

            _graphScores[graphId] =
                ((_graphScores[graphId] ?? 0) + scoreIncrease)
                    .clamp(0.0, 100.0)
                    .toDouble();
            _segmentsSinceLastTrigger[graphId] = 0;

            // --- Session timeline tracking ---
            _transitionHistory.add(TransitionRecord(
              graphId: graphId,
              fromStateId: fromStateId,
              toStateId: transition.targetStateId,
              scoreAfter: _graphScores[graphId]!,
              stage: targetState.stage,
              segmentIndex: _segmentIndex,
            ));
          }
          break;
        }
      }
    }

    // --- Cross-graph correlation bonus ---
    _applyCrossGraphBonus();

    _updateActiveScenario();
    return currentRiskScore;
  }

  double get currentRiskScore {
    var maxScore = 0.0;
    for (final score in _graphScores.values) {
      if (score > maxScore) maxScore = score;
    }
    return maxScore;
  }

  /// Minimum score floor: once decay brings score below this, it resets to 0.
  static const double _minScoreFloor = 2.0;

  /// Intent boost multiplier when both phrase + intent match a transition.
  static const double _intentBoostFactor = 1.5;

  /// Cross-graph bonus per additional active graph (2+ graphs = compound scam).
  static const double _crossGraphBonusPerGraph = 5.0;

  /// Minimum score to count a graph as "active" for cross-graph correlation.
  static const double _crossGraphActiveThreshold = 10.0;

  /// Adaptive decay: at higher stages (3-4), decay is slower to preserve
  /// strong scam signals. At stage 1, decay is faster to reduce stale signals.
  double _adaptiveDecay(double baseDecay, ScamStage stage) {
    return switch (stage) {
      ScamStage.stage1Introduction => baseDecay * 0.98, // faster decay
      ScamStage.stage2BaitingThreat => baseDecay, // normal decay
      ScamStage.stage3Urgency => baseDecay * 1.003, // slower decay
      ScamStage.stage4Command => baseDecay * 1.005, // very slow decay
    };
  }

  /// If 2+ graphs are active (score > threshold), apply compound scam bonus.
  void _applyCrossGraphBonus() {
    final activeCount = _graphScores.values
        .where((s) => s >= _crossGraphActiveThreshold)
        .length;
    if (activeCount >= 2) {
      final bonus = (activeCount - 1) * _crossGraphBonusPerGraph;
      // Apply bonus to the graph with the highest score
      MapEntry<String, double>? best;
      for (final entry in _graphScores.entries) {
        if (best == null || entry.value > best.value) {
          best = entry;
        }
      }
      if (best != null && best.value > 0) {
        _graphScores[best.key] =
            (best.value + bonus).clamp(0.0, 100.0).toDouble();
      }
    }
  }

  double _decayForGraph(String graphId) {
    // High-danger scenarios: very slow decay
    if (graphId.startsWith('G_POLICE') ||
        graphId.startsWith('G_KIDNAP') ||
        graphId.startsWith('G_SEXTORT') ||
        graphId.startsWith('G_BANK')) {
      return 0.995;
    }
    // Medium-danger scenarios
    if (graphId.startsWith('G_TECH') ||
        graphId.startsWith('G_HOSPITAL') ||
        graphId.startsWith('G_VNEID') ||
        graphId.startsWith('G_TELECOM') ||
        graphId.startsWith('G_CEO') ||
        graphId.startsWith('G_DEEPFAKE') ||
        graphId.startsWith('G_CREDIT') ||
        graphId.startsWith('G_CRYPTO')) {
      return 0.99;
    }
    // Low-danger / generic scenarios
    if (graphId.startsWith('G_CHARITY') ||
        graphId.startsWith('G_GENERIC') ||
        graphId.startsWith('G_GAMBLE') ||
        graphId.startsWith('G_ECOMMERCE')) {
      return 0.97;
    }
    return 0.98;
  }

  void _updateActiveScenario() {
    MapEntry<String, double>? bestEntry;
    for (final entry in _graphScores.entries) {
      if (bestEntry == null || entry.value > bestEntry.value) {
        bestEntry = entry;
      }
    }

    if (bestEntry == null || bestEntry.value <= 0) {
      activeScenarioName = null;
      activeScenarioStage = null;
      return;
    }

    ScenarioGraph? graph;
    for (final candidate in graphs) {
      if (candidate.graphId == bestEntry.key) {
        graph = candidate;
        break;
      }
    }
    activeScenarioName = graph?.name;
    final currentStateId = _currentSessionStates[bestEntry.key];
    final stateNode = graph?.states[currentStateId];
    activeScenarioStage = stateNode == null ? null : stateNode.stage.index + 1;
  }
}
